Fix initial pose bugs

12tbuchman opened the pull request:
Check if the odometry pose has been initialized and get the odometry pose if not.

When a new pose is sent to amcl from rviz or fleet, publish the pose and save it to the savedpose file.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33)
<!-- Reviewable:end -->

12tbuchman committed:
Fix initial pose bug

Check if the odometry pose has been initialized and get the odometry
pose if not.

bobhenz-jabil reviewed (changes requested):
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33)* status: all files reviewed, 4 unresolved discussions (waiting on @12tbuchman)

___
*[`-- commits` line 8 at r1](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33#-OLAStmu96ozGaiGdKe9:-OLAStmu96ozGaiGdKeA:b-82vyrt):*
This could probably easily (and should be) split into two separate commits.
___
*[`src/amcl/node/node.cpp` line 876 at r1](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33#-OLAT5Ud86te_tLv8fEg:-OLAT5Ud86te_tLv8fEh:b-pd58th) ([raw file](https://github.com/BadgerTechnologies/badger_amcl/blob/9784d8e9b0626ba519eaf853c17d9d6058314556/src/amcl/node/node.cpp#L876)):*
> ```cpp
>   {
>     Eigen::Vector3d pose;
>     getOdomPose(ros::Time::now(), &pose);
> ```

Should you check the return value of `getOdomPose()`? (Though I'm not sure what you would set last_pose to if that failed. :thinking: )
___
*[`src/amcl/node/node.cpp` line 911 at r1](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33#-OLAUjhk8WH25NYOu3I4:-OLAUjhk8WH25NYOu3I5:bz61ro1) ([raw file](https://github.com/BadgerTechnologies/badger_amcl/blob/9784d8e9b0626ba519eaf853c17d9d6058314556/src/amcl/node/node.cpp#L911)):*
> ```cpp
>   pose.getBasis().getRPY(roll, pitch, yaw);
>   Eigen::Vector3d init_pose(pose.getOrigin().x(), pose.getOrigin().y(), yaw);
>   publishPose(init_pose, ros::Time::now());
> ```

Should you use "ros::Time::now" or the time of the stamp from the message header?
___
*[`src/amcl/node/node.cpp` line 875 at r1](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33#-OLAUSFD2XbKdXdPYwh7:-OLAUSFD2XbKdXdPYwh8:boj6h8c) ([raw file](https://github.com/BadgerTechnologies/badger_amcl/blob/12b350abffba052b8dc70a4050af8c562cca6e44/src/amcl/node/node.cpp#L875)):*
> ```cpp
>   tf2::Transform map_pose = latest_tf.inverse() * latest_odom_pose_;
>   double yaw, pitch, roll;
>   map_pose.getBasis().getEulerYPR(yaw, pitch, roll);
> ```

Was this just unused? Is that why you were able to remove these lines?


<!-- Sent from Reviewable.io -->

12tbuchman reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33)* status: all files reviewed, 4 unresolved discussions (waiting on @bobhenz-jabil)

___
*[`src/amcl/node/node.cpp` line 875 at r1](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33#-OLAUSFD2XbKdXdPYwh7:-OLAZDut-wHflYjGkDp0:b2nign6) ([raw file](https://github.com/BadgerTechnologies/badger_amcl/blob/12b350abffba052b8dc70a4050af8c562cca6e44/src/amcl/node/node.cpp#L875)):*
<details><summary><i>Previously, bobhenz-jabil (Bob Henz) wrote…</i></summary><blockquote>

Was this just unused? Is that why you were able to remove these lines?
</blockquote></details>

yes
___
*[`src/amcl/node/node.cpp` line 876 at r1](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33#-OLAT5Ud86te_tLv8fEg:-OLAZBW4Bm45-SfEZOq3:b-896fix) ([raw file](https://github.com/BadgerTechnologies/badger_amcl/blob/9784d8e9b0626ba519eaf853c17d9d6058314556/src/amcl/node/node.cpp#L876)):*
<details><summary><i>Previously, bobhenz-jabil (Bob Henz) wrote…</i></summary><blockquote>

Should you check the return value of `getOdomPose()`? (Though I'm not sure what you would set last_pose to if that failed. :thinking: )
</blockquote></details>

Done.
___
*[`src/amcl/node/node.cpp` line 911 at r1](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33#-OLAUjhk8WH25NYOu3I4:-OLAZOmR-ElsbQuF-jnB:bwuwg65) ([raw file](https://github.com/BadgerTechnologies/badger_amcl/blob/9784d8e9b0626ba519eaf853c17d9d6058314556/src/amcl/node/node.cpp#L911)):*
<details><summary><i>Previously, bobhenz-jabil (Bob Henz) wrote…</i></summary><blockquote>

Should you use "ros::Time::now" or the time of the stamp from the message header?
</blockquote></details>

we dont have a message to get a header from


<!-- Sent from Reviewable.io -->

12tbuchman committed:
Fix initial pose received bug

When a new pose is sent to amcl from rviz or fleet, publish the pose and
save it to the savedpose file.

12tbuchman reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @bobhenz-jabil)

___
*[`-- commits` line 8 at r1](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33#-OLAStmu96ozGaiGdKe9:-OLAmY4g9nrA0aXUQmef:b-896fix):*
<details><summary><i>Previously, bobhenz-jabil (Bob Henz) wrote…</i></summary><blockquote>

This could probably easily (and should be) split into two separate commits.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

bobhenz-jabil reviewed (changes requested):
> Reviewed 1 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33)* status: 1 of 2 files reviewed, 2 unresolved discussions (waiting on @12tbuchman)

___
*[`src/amcl/node/node.cpp` line 911 at r1](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33#-OLAUjhk8WH25NYOu3I4:-OLAoZOhBLinaeAZ1857:b-59wji2) ([raw file](https://github.com/BadgerTechnologies/badger_amcl/blob/9784d8e9b0626ba519eaf853c17d9d6058314556/src/amcl/node/node.cpp#L911)):*
<details><summary><i>Previously, 12tbuchman (Tyler Buchman) wrote…</i></summary><blockquote>

we dont have a message to get a header from
</blockquote></details>

:point_up: 
```
const geometry_msgs::PoseWithCovarianceStampedConstPtr& msg_ptr
```
___
*[`src/amcl/node/node.cpp` line 386 at r2](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33#-OLAn3O04xjZmaWJkqbk:-OLAn3O04xjZmaWJkqbl:ba4bl5d) ([raw file](https://github.com/BadgerTechnologies/badger_amcl/blob/fee9b8eefd1cf78aa5ee8472308516792d7e57fe/src/amcl/node/node.cpp#L386)):*
> ```cpp
>     if (!getLatestPose(latest_tf, &latest_pose))
>     {
>       ROS_WARN("Failed to save pose. Could not get latest pose in odometry frame.");
> ```

I think you need a "return" statement after kicking out this warning or else won't we corrupt the saved pose?


<!-- Sent from Reviewable.io -->

bobhenz-jabil reviewed (commented):
> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33)* status: all files reviewed, 2 unresolved discussions (waiting on @12tbuchman)



<!-- Sent from Reviewable.io -->

12tbuchman reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33)* status: 1 of 2 files reviewed, 2 unresolved discussions (waiting on @bobhenz-jabil)

___
*[`src/amcl/node/node.cpp` line 911 at r1](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33#-OLAUjhk8WH25NYOu3I4:-OLApO3EAcnlxkwHt7vJ:b-tr9w1x) ([raw file](https://github.com/BadgerTechnologies/badger_amcl/blob/9784d8e9b0626ba519eaf853c17d9d6058314556/src/amcl/node/node.cpp#L911)):*
<details><summary><i>Previously, bobhenz-jabil (Bob Henz) wrote…</i></summary><blockquote>

:point_up: 
```
const geometry_msgs::PoseWithCovarianceStampedConstPtr& msg_ptr
```
</blockquote></details>

ah sorry, i thought this was in a different function
___
*[`src/amcl/node/node.cpp` line 386 at r2](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33#-OLAn3O04xjZmaWJkqbk:-OLApEF61lJ4T5JlkJWh:b-uq1cwc) ([raw file](https://github.com/BadgerTechnologies/badger_amcl/blob/fee9b8eefd1cf78aa5ee8472308516792d7e57fe/src/amcl/node/node.cpp#L386)):*
<details><summary><i>Previously, bobhenz-jabil (Bob Henz) wrote…</i></summary><blockquote>

I think you need a "return" statement after kicking out this warning or else won't we corrupt the saved pose?
</blockquote></details>

right, thanks


<!-- Sent from Reviewable.io -->

bobhenz-jabil reviewed (approved):
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BadgerTechnologies/badger_amcl/33)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @12tbuchman)



<!-- Sent from Reviewable.io -->
